platform/consul/monitor.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package consul
+
+import (
+	"reflect"
+	"sort"
+	"time"
+
+	"github.com/golang/glog"
+
+	"github.com/hashicorp/consul/api"
+	"istio.io/pilot/model"
+)
+
+type consulServices map[string][]string
+type consulServiceInstances []*api.CatalogService
+
+// Handler specifies a function to apply on an object for a given event type
+type Handler func(obj interface{}, event model.Event) error
+
+// Monitor handles service and instance changes
+type Monitor interface {
+	Start(<-chan struct{})
+	Stop()
```

Comment:
can you please document the concurrency requirements? can these two functions be called from two separate threads? what's the purpose of stop() is the stop channel can be used.